File: src/main/java/com/thesilentnights/easylogin/service/TaskService.java
package com.thesilentnights.easylogin.service;

import com.thesilentnights.easylogin.service.task.Loop;
import com.thesilentnights.easylogin.service.task.Task;

import java.util.TreeSet;
import java.util.UUID;

public class TaskService {
    static TreeSet<Task> tasks = new TreeSet<>();
    static int nextHook = 0;
    static int timePassedAfterPrepare = 0;

    public static void addTask(Task task) {
        reduceTickDelay();
        tasks.add(task);
        prepareNextTask();
    }

    public static void cancelPlayer(UUID uuid) {
        tasks.removeIf(task -> task.shouldCancel(uuid));
    }

    private static void reduceTickDelay() {
        tasks.forEach(task -> task.reduceTickDelay(timePassedAfterPrepare));
    }

    private static void prepareNextTask() {
        timePassedAfterPrepare = 0;
        if (tasks.isEmpty()) {
            return;
        }
        nextHook = tasks.first().getTickDelay();
    }

    public static void tick() {
        if (tasks.isEmpty()) {
            return;
        }
        timePassedAfterPrepare++;
        if (timePassedAfterPrepare > nextHook) {
            Task task = tasks.pollFirst();
            //tasks is not empty
            assert task != null;
            task.execute();

            //if is Loop task readd
            if (task instanceof Loop loop) {
                addTask(loop.regenerate());
            } else {
                reduceTickDelay();
                prepareNextTask();
            }
        }
    }


}

File: src/main/kotlin/com/thesilentnights/easylogin/service/TaskService.kt
package com.thesilentnights.easylogin.service

import com.thesilentnights.easylogin.service.task.Task

object TaskService {
    private val taskMap: MutableMap<kotlin.String, Task> = HashMap<kotlin.String, Task>()

    fun addTask(identifier: kotlin.String, task: Task) {
        taskMap[identifier] = task
    }

    fun getTask(identifier: kotlin.String): Task? {
        return taskMap.get(identifier)
    }

    fun generateTaskIdentifier(uuid: java.util.UUID, taskType: TaskService.TaskType): kotlin.String {
        return uuid.toString() + "_" + taskType
    }

    fun tick() {
        if (taskMap.isEmpty()) {
            return
        }
        taskMap.values.forEach({ task -> task.tick() })
    }

    fun cancelTask(identifier: kotlin.String?) {
        taskMap.remove(identifier)
    }

    fun cancelPlayer(uuid: java.util.UUID) {
        TaskService.taskMap.entries.removeIf { entry: kotlin.collections.MutableMap.MutableEntry<kotlin.String, Task> ->
            entry.key.startsWith(
                uuid.toString()
            )
        }
    }

    enum class TaskType {
        KICK_PLAYER, MESSAGE
    }
}